src/npyffi/array.rs

```diff
@@ -1,361 +1,338 @@
-/// Some APIs (including most accessor) in the above URL are not implemented
-/// since they are defined as C macro, and cannot be called from rust.
-/// Some of these are implemented on the high-level interface as a Rust function.
-pub struct PyArrayModule<'py> {
-    numpy: &'py PyModule,
-    api: *const *const c_void,
-}
-
-#[allow(non_upper_case_globals)]
-pub(crate) static mut PyArray_Type_Ptr: *mut PyTypeObject = null_mut();
+pub(crate) const MOD_NAME: &str = "numpy.core.multiarray";
+const CAPSULE_NAME: &str = "_ARRAY_API";
 
-impl<'py> Deref for PyArrayModule<'py> {
-    type Target = PyModule;
-    fn deref(&self) -> &Self::Target {
-        self.numpy
-    }
-}
+/// A global variable which stores a ['capsule'](https://docs.python.org/3/c-api/capsule.html)
+/// pointer to [Numpy Array API](https://docs.scipy.org/doc/numpy/reference/c-api.array.html).
+///
+/// You can acceess raw c APIs via this variable and its Deref implementation.
+///
+/// See [PyArrayAPI](struct.PyArrayAPI.html) for what methods you can use via this variable.
+///
+/// # Example
+/// ```
+/// # extern crate pyo3; extern crate numpy; fn main() {
+/// use numpy::{PyArray, npyffi::types::NPY_SORTKIND, PY_ARRAY_API};
+/// use pyo3::Python;
+/// let gil = Python::acquire_gil();
+/// let array: &PyArray<i32> = PyArray::arange(gil.python(), 2.0, 5.0, 1.0);
+/// array.as_slice_mut().unwrap().swap(0, 1);
+/// assert_eq!(array.as_slice().unwrap(), &[3, 2, 4]);
+/// unsafe {
+///     PY_ARRAY_API.PyArray_Sort(array.as_array_ptr(), 0, NPY_SORTKIND::NPY_QUICKSORT);
+/// }
+/// assert_eq!(array.as_slice().unwrap(), &[2, 3, 4])
+/// # }
+/// ```
+pub static PY_ARRAY_API: PyArrayAPI = PyArrayAPI {
+    __private_field: (),
+};
 
-/// Define Array API in PyArrayModule
-macro_rules! pyarray_api {
-    [ $offset:expr; $fname:ident ( $($arg:ident : $t:ty),* ) $( -> $ret:ty )* ] => {
-#[allow(non_snake_case)]
-pub unsafe fn $fname(&self, $($arg : $t), *) $( -> $ret )* {
-    let fptr = self.api.offset($offset) as (*const extern fn ($($arg : $t), *) $( -> $ret )* );
-    (*fptr)($($arg), *)
+///
+pub struct PyArrayAPI {
+    __private_field: (),
 }
-}} // pyarray_api!
 
-impl<'py> PyArrayModule<'py> {
-    /// Import `numpy.core.multiarray` to use Array API.
-    pub fn import(py: Python<'py>) -> PyResult<Self> {
-        let numpy = py.import("numpy.core.multiarray")?;
-        let c_api = numpy.getattr("_ARRAY_API")?;
-        let api = unsafe {
-            ffi::PyCapsule_GetPointer(c_api.as_ptr(), null_mut()) as *const *const c_void
-        };
-        let mod_ = PyArrayModule { numpy, api };
+impl Deref for PyArrayAPI {
+    type Target = PyArrayAPI_Inner;
+    fn deref(&self) -> &Self::Target {
+        static mut ARRAY_API_CACHE: PyArrayAPI_Inner = PyArrayAPI_Inner(ptr::null());
         unsafe {
-            PyArray_Type_Ptr = mod_.get_type_object(ArrayType::PyArray_Type);
+            // TODO: this operation is 'mostly safe' because of GIL, but not completely thread safe
```

Comment:
Do you have any plan to fix it?
I try to use lazy_static and Mutex, but it (https://github.com/rust-numpy/rust-numpy/commit/9bd437690a35fbd57113dad1f69148d91671485f) does not work :<

Reply:
Somewhat calling python ffi from synchronized path causes dead lock when `cargo test` executes multiple tests in parallel.
But I think we can ensure safety of substitution to ARRAY_API_CACHE by `std::sync::Once`

Reply:
i.e.
This is OK
```rust
let api = get_numpy_api(MOD_NAME, CAPSULE_NAME);
INIT_API.call_once(move || {
    ARRAY_API_CACHE = PyArrayAPI_Inner(api);
});
```
but this causes dead lock, in parallel tests
```rust
INIT_API.call_once(move || {
    let api = get_numpy_api(MOD_NAME, CAPSULE_NAME);
    ARRAY_API_CACHE = PyArrayAPI_Inner(api);
});
```

Reply:
And I think we don't have to use `Mutex`.
We have to lock the access to Numpy API only once in a program, since Numpy API is pinned by capsule.

Reply:
It seems to be correct.